src/main/java/org/veupathdb/eda/binaryfiles/dumper/StudyDumper.java

```diff
@@ -42,33 +44,48 @@ public void dumpStudy() {
     Entity rootEntity = entityTree.getContents();
     
     // Root study gets a special IDs file dumper (doesn't need parent ancestors)
-    dumpSubtree(entityTree, () -> new IdFilesDumperNoAncestor(_bfm, _study, rootEntity));
+    dumpSubtree(entityTree, new IdFilesDumperFactory(_bfm, _study, rootEntity, null), new HashMap<>());
     writeDoneFile(_bfm.getStudyDir(_study, Operation.READ));
   }
 
-  private void dumpSubtree(TreeNode<Entity> subTree, Supplier<FilesDumper> idsDumperSupplier) {
+  private void dumpSubtree(TreeNode<Entity> subTree, IdFilesDumperFactory idDumperFactory, Map<String, Integer> entityIdToMaxIdLength) {
     Entity entity = subTree.getContents();
-    
-    dumpEntity(entity, idsDumperSupplier);
+
+    // Find the maximum length of an ID for this entity to determine space to allocate for each ID.
+    // Add 4 to store the size of the ID as well.
+    int maxIdLength = scanForMaxIdLength(entity);
+
+    // Add the value to the Map so that it is available when dumping subtrees recursively.
+    Map<String, Integer> newMaxIdLengthMap = new HashMap(entityIdToMaxIdLength);
```

Comment:
Why make a copy here?  Cheap, but necessary?  The entities inside will be added top-down so I could see only wanting the actual ancestors of the "current" entity, but then you loop over the ancestors below at 79 (not use e.g. new JSONObject(entityIdToMaxLength)).  Just trying to understand- maybe a comment as to why a copy is preferred?

Reply:
Not necessary, I think I just did it out of paranoia about the side effect even though it shouldn't be an issue. I'll avoid the copy to make it less confusing